enhancements/builds/volume-secrets.md

```diff
@@ -0,0 +1,129 @@
+---
+title: volume-mounted-secret-and-configmap-injection
+authors:
+  - "@bparees"
+reviewers:
+  - @adambkaplan
+  - @smarterclayton
+approvers:
+  - @adambkaplan
+creation-date: 2020-01-08
+last-updated: 2020-01-08
+status: provisional
+---
+
+# Volume Mounted Secret+ConfigMap Injections
+
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+ > 1. Can we make this the default or even only behavior for builds?  Probably not, need to make it opt-in to avoid potentially breaking existing buildconfigs.
+
+## Summary
+
+Today builds support getting source input from configmaps and secrets.  When users utilize this feature, the configmap or secret is volume-mounted into the build pod.  The next steps depend on whether it is an s2i or dockerfile build.
+
+For s2i builds, the generated Dockerfile contains commands to `ADD` the content at a path specified by the user, the assemble script is invoked, and then the injected content is zeroed out prior to committing the image via a `RUN rm` command added to the Dockerfile.
+
+For dockerfile builds, the user is instructed to add appropriate `ADD` and `RUN rm` commands to their dockerfile to inject the content that is available in the build's working directory (along with their application source, where applicable).
+
+There are a few undesirable aspects to this:
+1) In the dockerfile case, the content can still be found in lower layers of the image unless a layer squashing option is 
+selected.
+2) Requires extra work by the user in the Dockerfile, so each Dockerfile must be customized
+
+This enhancement proposes to use buildah's capability to mount a volume at build time.  The content mounted into the build pod would be then mounted into the container processing the Dockerfile, making that content available within the container so Dockerfile commands could reference it.  No explicit `ADD` would be required, and since mounted content is not committed to the resulting image, no `RUN rm` equivalent is required to clean up the injected content.
+
+
+## Motivation
+
+### Goals
+
+* Simplify how users consume secret + configmap content in builds
+* Increase the security of protected content being injected to images
+* Simplify use cases like mounting subscription entitlement credentials into a build so subscription content can be pulled down during a build while not including the credentials in the resulting image.
+
+### Non-Goals
+
+* This enhancement is not related to the frequently requested "build volumes" feature which would make it possible to mount a volume during a build and read/write content on that volume during the build such as to cache dependencies.
```

Comment:
Hrm - I kind of expected you to say “we should not create a solution for mounting multiple volume types in different ways, so while we aren’t solving for that we expect to expose an api compatible with the future”.

Ie I expect mounting secrets for pods and builds to be the same api experience.

Reply:
well that ship already sailed to some degree (in terms of build-pod volume mounts.  not in terms of buildah-mounting).  We are always going to have the existing secret/configmap input api which controls build-pod mounts.

we may also have the api you seem to be asking for, which would be a secondary way to drive build-pod mounts.

but users don't care about build-pod mounts since that has no real impact on experience.  they care about whether we are doing buildah-mounts, or injecting the content in a different way, since that determines what ultimately ends up in the output image.

in any case your suggestion has benefits in terms of pushing us towards the "support PVs" direction, though at the short term cost of being a larger implementation for the immediate goal of treating secrets as a buildah-mount.